content/client-lib-development-guide/features.textile

```diff
@@ -599,6 +599,16 @@ h3(#realtime-channel). Channel
 ** @(RTL7d)@ Messages delivered are automatically decoded based on the @encoding@ attribute; see REST @Channel@ encoding features. Tests should exist to publish and subscribe to encoded messages using the "AES 128":https://github.com/ably/ably-common/blob/main/test-resources/crypto-data-128.json and "AES 256":https://github.com/ably/ably-common/blob/main/test-resources/crypto-data-256.json fixture test data
 ** @(RTL7e)@ If a message cannot be decoded or decrypted successfully, it should be delivered to the listener with the @encoding@ attribute set indicating the residual encoding state, and an error should be logged
 ** @(RTL7f)@ A test should exist ensuring published messages are not echoed back to the subscriber when @echoMessages@ is set to false in the @Realtime@ library constructor
+** @(RTL7g)@ Subscribe with a dictionary argument subscribes a listener which filters which messages to accept based on the dictionary values
+*** @(RTL7g1)@ An empty dictionary functions identically to "RTL7a":#RTL7a
+*** @(RTL7g2)@ A dictionary with multiple key/value pairs should require each entry to be satisfied in order to accept the Message
+*** @(RTL7g3)@ Keys currently must be one of @ref@ or @type@ or a negated version of these (@-ref@ and @-type@)
+**** @(RTL7g3a)@ @ref@ and @type@ refer to @ref.timeserial@ and @ref.type@ in the Message body respectively
+**** @(RTL7g3b)@ Consideration should be made to allow for easy expansion of the allowed fields here in the future 
+*** @(RTL7g4)@ Keys which are negated (starting with a @-@) should be treated as a NOT and not accept any Messages which match that value
+*** @(RTL7g5)@ A value should be matched exactly, unless it is @*@ which should match any value that is not null
```

Comment:
I don't think this is needed, if you leave the field blank you get any value and I've not seen evidence for needing to support more complex regexes.

If we think that if more complex filters are needed then we should us a functional argument instead e.g. `filter(msg): bool`. The benefit I see in using exact strings is that you can put the listeners in a hashmap and have more efficient lookups on a per message basis. However if/when the dictionary approach is extended to support new fields it will be a bit of a nightmare of hierarchies.

If we assume that there will be <5 handlers per message in view in the chat and <20 messages in view at a time then a new inbound message will call < 400 filter functions. It's not amazing but it is definitely the most generally useful.

So I think we should go with a function rather than dict. WDYT?

Reply:
>  if you leave the field blank you get any value and I've not seen evidence for needing to support more complex regexes.  
The asterisk's purpose is mostly for being able to handle only messages with no ref, e.g `"-ref": "*"`. I felt that was more extensible than the `includeRefs` we drafted. A function would be much more flexible and definitely easier to implement so I'm definitely happy to go that route instead. Would we still have the dictionary version for simple filtering or would be only use the function?